tests/app/main/test_suppliers.py

```diff
@@ -1665,3 +1667,223 @@ def test_should_show_email_address_even_when_refreshed(self):
+
+            assert doc.xpath(
+                "//*[contains(@class, 'banner-destructive-without-action')][contains(normalize-space(string()), $t)]//"
+                "a[@href=$m][normalize-space(string())=$e]",
+                t="The service is unavailable at the moment",
+                m="mailto:enquiries@digitalmarketplace.service.gov.uk",
+                e="enquiries@digitalmarketplace.service.gov.uk",
+            )
+
+            with c.session_transaction() as session:
+                # flash message should have been consumed by view's own page rendering
+                assert not session.get('_flashes')
+
+    @mock.patch("app.main.views.suppliers.DMMailChimpClient")
+    def test_post_valid_email_success(self, mailchimp_client_class, data_api_client):
+        data_api_client.create_audit_event.side_effect = assert_args_and_return(
+            {"convincing": "response"},
+            audit_type=AuditTypes.mailing_list_subscription,
+            data={
+                "subscribedEmail": "qu&rt@four.pence",
+                "mailchimp": {
+                    "id": "cashregister-clanged",
+                    "unique_email_id": "clock-clacked",
+                    "timestamp_opt": None,
+                    "last_changed": "1904-06-16T16:00:00+00:00",
+                    "list_id": "flowered-tables",
+                },
+            },
+        )
+        mailchimp_client_instance = mock.Mock(spec=("subscribe_new_email_to_list",))
+        mailchimp_client_instance.subscribe_new_email_to_list.side_effect = assert_args_and_return(
+            {
+                "id": "cashregister-clanged",
+                "unique_email_id": "clock-clacked",
+                # timestamp_opt deliberately omitted
+                "last_changed": "1904-06-16T16:00:00+00:00",
+                "list_id": "flowered-tables",
+                "has-he-forgotten": "perhaps-a-trick",  # should be ignored
+            },
+            "not_a_real_mailing_list",
+            "qu&rt@four.pence",
+        )
+
+        mailchimp_client_class.side_effect = assert_args_and_return(
+            mailchimp_client_instance,
+            "not_a_real_username",
+            "not_a_real_key",
+            mock.ANY,
+        )
+
+        with self.client as c:
+            response = c.post(
+                "/suppliers/mailing-list",
+                data={
+                    # no, ampersands are probably not valid in this position in an email address but they are accepted
+                    # by our regex and we want to be able to check their escaping in the flash message
+                    "email_address": "qu&rt@four.pence",
+                },
+            )
+
+            assert response.status_code == 302
+            assert response.location == "http://localhost/"
+            assert mailchimp_client_instance.subscribe_new_email_to_list.called is True
+
+            with c.session_transaction() as session:
```

Comment:
I think this is the same thing that the helper `self.assert_flashes` does.  Might be nice to stick with using that.

Reply:
Done. Interesting that `assert_flashes` doesn't seem to give a crap about binding to the same client context, so that casts questions on the use of any of the `with self.client as c:`, so I've removed that too.